fivox/densityImageSource.h

```diff
@@ -0,0 +1,62 @@
+
+/* Copyright (c) 2014-2016, EPFL/Blue Brain Project
+ *                          Daniel.Nachbaur@epfl.ch
+ *
+ * This file is part of Fivox <https://github.com/BlueBrain/Fivox>
+ *
+ * This library is free software; you can redistribute it and/or modify it under
+ * the terms of the GNU Lesser General Public License version 3.0 as published
+ * by the Free Software Foundation.
+ *
+ * This library is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
+ * FOR A PARTICULAR PURPOSE.  See the GNU Lesser General Public License for more
+ * details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this library; if not, write to the Free Software Foundation, Inc.,
+ * 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
+ */
+
+#ifndef FIVOX_DENSITYIMAGESOURCE_H
+#define FIVOX_DENSITYIMAGESOURCE_H
+
+#include <fivox/imageSource.h>
+#include <fivox/types.h>
+
+namespace fivox
+{
+
+/** Image source iterating over all events to calculate density per voxel. */
+template< typename TImage >
+class DensityImageSource : public ImageSource< TImage >
+{
+public:
+    /** Standard class typedefs. */
+    typedef DensityImageSource              Self;
+    typedef ImageSource< TImage >           Superclass;
+    typedef itk::SmartPointer< Self >       Pointer;
+    typedef itk::SmartPointer< const Self > ConstPointer;
+
+    /** Method for creation through the object factory. */
+    itkNewMacro(Self)
+
+    /** Run-time type information (and related methods). */
+    itkTypeMacro(DensityImageSource, ImageSource)
+
+protected:
+    DensityImageSource();
+    virtual ~DensityImageSource() {}
+
+    void GenerateData() override;
```

Comment:
GenerateData() is non-threaded, ThreadedGenerateData() is threaded. That's ITK. Not sure what happens if you override both...